src/main/kotlin/fi/elsapalvelu/elsa/service/impl/ArviointityokaluServiceImpl.kt

```diff
@@ -1,29 +1,119 @@
 package fi.elsapalvelu.elsa.service.impl
 
+import fi.elsapalvelu.elsa.domain.ArviointityokaluKysymys
+import fi.elsapalvelu.elsa.domain.ArviointityokaluKysymysVaihtoehto
+import fi.elsapalvelu.elsa.domain.AsiakirjaData
+import fi.elsapalvelu.elsa.repository.ArviointityokaluKategoriaRepository
 import fi.elsapalvelu.elsa.repository.ArviointityokaluRepository
+import fi.elsapalvelu.elsa.repository.KayttajaRepository
 import fi.elsapalvelu.elsa.service.ArviointityokaluService
 import fi.elsapalvelu.elsa.service.dto.ArviointityokaluDTO
+import fi.elsapalvelu.elsa.service.dto.UserDTO
+import fi.elsapalvelu.elsa.service.mapper.ArviointityokaluKysymysMapper
+import fi.elsapalvelu.elsa.service.mapper.ArviointityokaluKysymysVaihtoehtoMapper
 import fi.elsapalvelu.elsa.service.mapper.ArviointityokaluMapper
 import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
+import org.springframework.web.multipart.MultipartFile
+import java.time.Instant
 import java.util.*
 
 @Service
 @Transactional
 class ArviointityokaluServiceImpl(
     private val arviointityokaluRepository: ArviointityokaluRepository,
-    private val arviointityokaluMapper: ArviointityokaluMapper
+    private val arviointityokaluMapper: ArviointityokaluMapper,
+    private val arviointityokaluKysymysMapper: ArviointityokaluKysymysMapper,
+    private val arviointityokaluKysymysVaihtoehtoMapper: ArviointityokaluKysymysVaihtoehtoMapper,
+    private val arviointityokaluKategoriaRepository: ArviointityokaluKategoriaRepository,
+    private val kayttajaRepository: KayttajaRepository,
 ) : ArviointityokaluService {
 
-    override fun save(arviointityokaluDTO: ArviointityokaluDTO): ArviointityokaluDTO {
+    override fun save(
+        arviointityokaluDTO: ArviointityokaluDTO,
+        user: UserDTO,
+        liiteData: MultipartFile?
+    ): ArviointityokaluDTO {
         var arviointityokalu = arviointityokaluMapper.toEntity(arviointityokaluDTO)
+        val now = Instant.now()
+        if (arviointityokalu.id == null) {
+            arviointityokalu.luontiaika = now
+        }
+        arviointityokalu.muokkausaika = now
+        arviointityokalu.kayttaja = user.id?.let { kayttajaRepository.findOneByUserId(it).orElse(null) }
+        arviointityokalu.kategoria = arviointityokalu.kategoria?.let {
+            arviointityokaluKategoriaRepository.findById(it.id!!).orElse(null)
+        }
+        arviointityokalu.kysymykset.forEach { kysymys ->
+            kysymys.arviointityokalu = arviointityokalu
+            kysymys.vaihtoehdot.forEach { vaihtoehto ->
+                vaihtoehto.arviointityokaluKysymys = kysymys
+            }
+        }
```

Comment:
Tarvitseekohan näitä määrittää itse? Muistelin että JPA osaisi tehdä nuo itsestään kun arviointityökalu tallennetaan.

Reply:
Meillä onkin muut tapaukset olleet hieman erilaisia ja mapstruct on hoitanut kaikki referenssit kohdilleen. Tähän tapaukseen ei suoraan löytynyt helppoa ratkaisua eli voi jättää näin